src/jsCodegen/emitTest.ts

```diff
@@ -0,0 +1,92 @@
+import { NODE_TYPES } from '../domFacade/ConcreteNode';
+import astHelper, { IAST } from '../parsing/astHelper';
+import { acceptAst, PartiallyCompiledJavaScriptResult, rejectAst } from './CompiledJavaScript';
+
+const testAstNodeNames = {
+	TEXT_TEST: 'textTest',
+	ELEMENT_TEST: 'elementTest',
+	NAME_TEST: 'nameTest',
+	WILDCARD: 'Wildcard',
+};
+
+export const testAstNodes = Object.values(testAstNodeNames);
+
+const testEmittersByAstNodeName = {
+	[testAstNodeNames.TEXT_TEST]: emitTextTest,
+	[testAstNodeNames.NAME_TEST]: emitNameTest,
+	[testAstNodeNames.ELEMENT_TEST]: emitElementTest,
+	[testAstNodeNames.WILDCARD]: emitWildcard,
+};
+
+// text() matches any text node.
+// https://www.w3.org/TR/xpath-31/#doc-xpath31-TextTest
+function emitTextTest(_ast: IAST, identifier: string): PartiallyCompiledJavaScriptResult {
+	return acceptAst(`${identifier}.nodeType === ${NODE_TYPES.TEXT_NODE}`);
+}
+
+type QName = { localName: string; namespaceURI: string; prefix: string };
+
+function emitNameTestFromQName(
+	identifier: string,
+	{ prefix, namespaceURI, localName }: QName
+): PartiallyCompiledJavaScriptResult {
+	const conditionCode = `${identifier}.nodeType && (${identifier}.nodeType === ${NODE_TYPES.ELEMENT_NODE} || ${identifier}.nodeType === ${NODE_TYPES.ATTRIBUTE_NODE})`;
+	if (namespaceURI) {
+		return rejectAst('Unsupported: namespace URI in name tests.');
+	}
+
+	if (prefix === '*' && localName === '*') {
+		return acceptAst(conditionCode);
+	}
+
+	return acceptAst(
+		`${conditionCode} && ${identifier}.localName === ${JSON.stringify(localName)}`
```

Comment:
aren't we missing a Namespace test here?

`self::p` should only match the `p` element when it is in the namespace the namespaceResolver resolves the empty prefix to